### PR TITLE
[SPARK-40142][PYTHON][SQL][FOLLOW-UP] Make pyspark.sql.functions examples self-contained (part 6, ~50 functions)

### PR DESCRIPTION
**Pull request by khalidmammadov**
### What changes were proposed in this pull request?
It's part of the Pyspark docstrings improvement series (https://github.com/apache/spark/pull/37592, https://github.com/apache/spark/pull/37662, https://github.com/apache/spark/pull/37686, https://github.com/apache/spark/pull/37786)

In this PR I mainly covered missing parts in the docstrings adding some more examples where it needed.

### Why are the changes needed?

To improve PySpark documentation

### Does this PR introduce _any_ user-facing change?

Yes, documentation

### How was this patch tested?
```
PYTHON_EXECUTABLE=python3.9 ./dev/lint-python
./python/run-tests --testnames pyspark.sql.functions
bundle exec jekyll build
```

### CODE REVIEW

**Comment by khalidmammadov** [comment]
@srowen @HyukjinKwon @itholic  please review

File: python/pyspark/sql/functions.py [hunk 1]
```diff
@@ -4662,29 +4735,36 @@ def xxhash64(*cols: "ColumnOrName") -> Column:
...
     >>> df.select(assert_true(df.a < df.b).alias('r')).collect()
     [Row(r=None)]
-    >>> df = spark.createDataFrame([(0,1)], ['a', 'b'])
     >>> df.select(assert_true(df.a < df.b, df.a).alias('r')).collect()
     [Row(r=None)]
-    >>> df = spark.createDataFrame([(0,1)], ['a', 'b'])
     >>> df.select(assert_true(df.a < df.b, 'error').alias('r')).collect()
     [Row(r=None)]
+    >>> df.select(assert_true(df.a > df.b, 'My error msg').alias('r')).collect() # doctest: +SKIP
+    22/09/03 20:18:15 ERROR Executor: Exception in task 15.0 in stage 45.0 (TID 383)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4765]: Let's probably replace this line to `...`

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -4707,6 +4787,19 @@ def raise_error(errMsg: Union[Column, str]) -> Column:
...
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        throws an error with specified message.
+
+    Examples
+    --------
+    >>> df = spark.range(1)
+    >>> df.select(raise_error("My error message")).show() # doctest: +SKIP
+    22/09/03 20:26:49 ERROR Executor: Exception in task 15.0 in stage 46.0 (TID 399)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4800]: ditto. probably let's replace this to `...`

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -4725,6 +4818,29 @@ def upper(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        upper case values.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4835]:
```suggestion
    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], "STRING")
```

nit but shorter

File: python/pyspark/sql/functions.py [hunk 4]
```diff
@@ -4734,6 +4850,29 @@ def lower(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        lower case values.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4867]:
```suggestion
    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -4743,6 +4882,29 @@ def ascii(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        numeric value.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4899]:
```suggestion
    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -4752,6 +4914,29 @@ def base64(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        BASE64 encoding of string value.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4931]:
```suggestion
    >>> df = spark.createDataFrame(["Spark", "PySpark", "Pandas API"], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -4761,6 +4946,31 @@ def unbase64(col: "ColumnOrName") -> Column:
...
+    -------
+    :class:`~pyspark.sql.Column`
+        encoded string value.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["U3Bhcms=",
+    ...                             "UHlTcGFyaw==",
+    ...                             "UGFuZGFzIEFQSQ=="], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4965]:
```suggestion
    >>> df = spark.createDataFrame(["U3Bhcms=",
    ...                             "UHlTcGFyaw==",
    ...                             "UGFuZGFzIEFQSQ=="], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 8]
```diff
@@ -4770,6 +4980,29 @@ def ltrim(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        left trimmed values.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:4997]:
```suggestion
    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 9]
```diff
@@ -4779,6 +5012,29 @@ def rtrim(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        right trimmed values.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:5029]:
```suggestion
    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], "STRING")
```

File: python/pyspark/sql/functions.py [hunk 10]
```diff
@@ -4788,6 +5044,29 @@ def trim(col: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        trimmed values from both sides.
+
+    Examples
+    --------
+    >>> from pyspark.sql import types
+    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], types.StringType())
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py:5061]:
```suggestion
    >>> df = spark.createDataFrame(["   Spark", "Spark  ", " Spark"], "STRING")
```

**Comment by HyukjinKwon** [comment]
Looks pretty good otherwise. cc @itholic @xinrong-meng @Yikun @zhengruifeng in case you guys find some time to review.

File: python/pyspark/sql/functions.py [hunk 11]
```diff
@@ -4638,6 +4691,16 @@ def hash(*cols: "ColumnOrName") -> Column:
 
     .. versionadded:: 2.0.0
 
+    Parameters
+    ----------
+    cols : :class:`~pyspark.sql.Column` or str
+        list of columns to work on.
```
Comment (itholic) [inline python/pyspark/sql/functions.py:4697]: nit: Maybe `column or list of columns to compute on` ??

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -4638,6 +4691,16 @@ def hash(*cols: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        hash value as int column.
+
     Examples
     --------
     >>> spark.createDataFrame([('ABC',)], ['a']).select(hash('a').alias('hash')).collect()
```
Comment (itholic) [inline python/pyspark/sql/functions.py:4706]: Can we also have one more example using list of columns ??

File: python/pyspark/sql/functions.py [hunk 13]
```diff
@@ -4652,6 +4715,16 @@ def xxhash64(*cols: "ColumnOrName") -> Column:
 
     .. versionadded:: 3.0.0
 
+    Parameters
+    ----------
+    cols : :class:`~pyspark.sql.Column` or str
+        list of columns to work on.
```
Comment (itholic) [inline python/pyspark/sql/functions.py:4721]: ditto ?

File: python/pyspark/sql/functions.py [hunk 14]
```diff
@@ -4652,6 +4715,16 @@ def xxhash64(*cols: "ColumnOrName") -> Column:
...
+
+    Returns
+    -------
+    :class:`~pyspark.sql.Column`
+        hash value as long column.
+
     Examples
     --------
     >>> spark.createDataFrame([('ABC',)], ['a']).select(xxhash64('a').alias('hash')).collect()
```
Comment (itholic) [inline python/pyspark/sql/functions.py:4730]: ditto ?

File: python/pyspark/sql/functions.py (hunk 11 above)
Comment (itholic) [reply python/pyspark/sql/functions.py:4697]: And maybe we should mention `list` for the available types ??

e.g.
```
cols : :class:`~pyspark.sql.Column`, list or str
```

**Review by itholic — COMMENTED** [review]
Looks good otherwise

**Review by HyukjinKwon — APPROVED** [review]
LGTM from a cursory look. @xinrong-meng would you mind taking a quick look (and probably merging) this one please? 🙏

**Review by Yikun — APPROVED** [review]
LGTM

**Review by itholic — APPROVED** [review]
LGTM !

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**